go/vt/wrangler/reparent.go

```diff
@@ -406,52 +408,191 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
 		if err != nil {
-			return fmt.Errorf("can't get current replication position of master: %v", err)
+			return vterrors.Wrapf(err, "failed to promote %v to master", masterElectTabletAliasStr)
 		}
 		reparentJournalPos = rp
-	} else {
-		// If the current master is not the one we want, demote the old master.
-		// Note that since the shard record is eventually consistent, we might be
-		// demoting a former master that was already demoted.
+	} else if topoproto.TabletAliasEqual(currentMaster.Alias, masterElectTabletAlias) {
+		refreshCtx, refreshCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+		defer refreshCancel()
+
+		// The master is already the one we want according to its tablet record.
+		// Refresh it to make sure the tablet has read its record recently.
+		if err := wr.tmc.RefreshState(refreshCtx, masterElectTabletInfo.Tablet); err != nil {
+			return vterrors.Wrapf(err, "failed to RefreshState on current master %v", masterElectTabletAliasStr)
+		}
 
-		oldMasterTabletInfo, ok := tabletMap[topoproto.TabletAliasString(shardInfo.MasterAlias)]
-		if !ok {
-			return fmt.Errorf("old master tablet %v is not in the shard", topoproto.TabletAliasString(shardInfo.MasterAlias))
+		// Then get the position so we can try to fix replicas (below).
+		rp, err := wr.tmc.MasterPosition(refreshCtx, masterElectTabletInfo.Tablet)
+		if err != nil {
+			return vterrors.Wrapf(err, "failed to get replication position of current master %v", masterElectTabletAliasStr)
 		}
+		reparentJournalPos = rp
+	} else {
+		// There is already a master and it's not the one we want.
+		oldMasterTabletInfo := currentMaster
 		ev.OldMaster = *oldMasterTabletInfo.Tablet
 
-		// Demote the old master and get its replication position.
-		wr.logger.Infof("demote current master %v", shardInfo.MasterAlias)
+		// Before demoting the old master, check replication on the candidate
+		// master to make sure it has a chance of catching up.
+		statusCtx, statusCancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+		defer statusCancel()
+		status, err := wr.tmc.SlaveStatus(statusCtx, masterElectTabletInfo.Tablet)
+		if err != nil {
+			return vterrors.Wrapf(err, "can't get replication status on master-elect %v", masterElectTabletAliasStr)
+		}
+		if !status.SlaveIoRunning || !status.SlaveSqlRunning {
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "replication not running on master-elect %v; replication must be healthy to perform planned reparent", masterElectTabletAliasStr)
+		}
+		// Check if it's behind by a small enough amount.
+		if float64(status.SecondsBehindMaster) > masterElectLagThreshold.Seconds() {
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "replication lag on master-elect %v (%v seconds) is greater than the specified lag threshold (%v); let replication catch up first or try again with a higher threshold", masterElectTabletAliasStr, status.SecondsBehindMaster, masterElectLagThreshold)
```

Comment:
lag threshold -> lag_threshold
to be consistent with user-visible flag.

Reply:
I removed this flag entirely since we don't check lag anymore, as discussed offline.